src/cursor-doc/token-cursor.ts

```diff
@@ -117,34 +117,29 @@ export class TokenCursor {
  * If you are particular about which list type that should be considered, supply an `openingBracket`.
  */
 
-function _rangesForSexpsInList(
+const collectRanges = <
+  StartFieldKey extends keyof LispTokenCursor,
+  EndFieldKey extends keyof LispTokenCursor
+>(
   cursor: LispTokenCursor,
-  useRowCol = false,
-  openingBracket?: string
-): [number, number][] | [[number, number], [number, number]][] {
-  if (openingBracket !== undefined) {
-    if (!cursor.backwardListOfType(openingBracket)) {
-      return undefined;
-    }
-  } else {
-    if (!cursor.backwardList()) {
-      return undefined;
-    }
-  }
-  const ranges = [];
+  cursorStartField: StartFieldKey,
+  cursorEndField: EndFieldKey
+): [LispTokenCursor[StartFieldKey], LispTokenCursor[EndFieldKey]][] => {
+  const ranges: [LispTokenCursor[StartFieldKey], LispTokenCursor[EndFieldKey]][] = [];
```

Comment:
The explanation for this function is that we needed a generic version of `_rangesForSexpsInList` that didn't need really long types and weird type contortions in order to use it.

Reply:
It's probably an improvement. But with my limited command over the typing system, right now this turned from ”this is messy” to ”I don't understand what I'm reading at all”. Not sure what to do about that...

Reply:
It might also be that I am having a hard time trying to puzzle together the code in my head from the diffs.